tsl/src/fdw/deparse.c

```diff
@@ -834,6 +834,169 @@ deparseSelectStmtForRel(StringInfo buf, PlannerInfo *root, RelOptInfo *rel, List
 	deparseLockingClause(&context);
 }
 
+/*
+ * Construct "SELECT DISTINCT target_list" or "SELECT DISTINCT ON (col1, col..)
+ * target_list" statement to push down the DISTINCT clause to the remote side.
+ *
+ * We only allow references to basic "Vars" or constants in the DISTINCT exprs
+ *
+ * So, "SELECT DISTINCT col1" is fine but "SELECT DISTINCT 2*col1" is not.
+ *
+ * "SELECT DISTINCT col1, 'const1', NULL, col2" which is a mix of column
+ * references and constants is also supported. Everything else is not supported.
+ *
+ * Note that "SELECT DISTINCT col1, col2" is different from "SELECT DISTINCT
+ * col2, col1". So, the attributes have to be fetched in the same order as
+ * well from the remote side. This means that "retrieved_attrs" list also
+ * needs to be constructed in the same order. This function returns a
+ * "dc_varattnos" list which points to the attributes as per the DISTINCT
+ * specification.
+ */
+static void
+deparseDistinctClause(StringInfo buf, List **dc_varattnos, deparse_expr_cxt *context)
+{
+	PlannerInfo *root = context->root;
+	Query *query = root->parse;
+	ListCell *l, *dc_l;
+	bool first = true;
+	bool dc_pushdown_safe = true;
+	bool have_wholerow = false;
+	Index varno;
+	/*
+	 * Sortgroups to deparse for SELECT DISTINCT. Only "Vars" will go in
+	 * this list. Constants will not go
+	 */
+	List *sgc_deparse_list = NIL;
+
+	*dc_varattnos = NIL;
+
+	if (query->distinctClause == NIL)
+		return;
```

Comment:
Seems appropriate to move this up before setting `dc_varattnos` to avoid side effects in case of no distinct clause.

Reply:
That is intentional. We want to the dc_varattnos to be explicitly `NIL` if there's no distinct clause.

Reply:
Then I think it is more appropriate that `dc_varattnos` is a return value rather than a passed in variable with unclear side-effects.